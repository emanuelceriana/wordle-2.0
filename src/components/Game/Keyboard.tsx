import { useCallback, useContext, useMemo, useState } from "react";
import { IAppContext, AppContext } from "../../context/AppContext";
import {
  faArrowTurnDown,
  faDeleteLeft,
  faCaretDown,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import cn from "classnames";
import { keyboardButtons, keyboardButtonsMobile } from "./constants";
import { useCreateObjectList } from "../hooks/useCreateObjectList";
import styles from "./Keyboard.module.scss";
import { isMobile } from "react-device-detect";

interface KeyboardProps {
  rowRefs: React.MutableRefObject<HTMLDivElement[]>;
  keyboardPreview: string[];
}

export const Keyboard = ({ rowRefs, keyboardPreview }: KeyboardProps) => {
  const [isHiddenKeyboard, setIsHiddenKeyboard] = useState(false);
  const { globalWordValidation, activeRowIdx } =
    useContext<IAppContext>(AppContext);

  const keyboard = useMemo(() => {
    return isMobile ? keyboardButtonsMobile : keyboardButtons;
  }, [isMobile]);

  const { objectList: keyboardButtonsIds } = useCreateObjectList({
    length: keyboard.length,
  });

  const handleClick = useCallback(
    (letter: string) => {
      const event = new KeyboardEvent("keyup", {
        key: letter,
        bubbles: true,
      });

      rowRefs.current[activeRowIdx].dispatchEvent(event);
    },
    [rowRefs, activeRowIdx]
  );

  return (
    <div
      className={cn(styles.keyboard, {
        [styles.isHiddenKeyboard]: isHiddenKeyboard,
      })}
    >
      {isMobile && (
        <>
          <div
            data-testid="hideMobileKeyboard"
            className={styles.mobileVisibilityControl}
            onClick={() => setIsHiddenKeyboard(!isHiddenKeyboard)}
          >
            <FontAwesomeIcon icon={faCaretDown} />
          </div>
          <div className={styles.keyboardPreview}>
            {keyboardPreview?.join("")}
          </div>
        </>
      )}
      {keyboardButtonsIds.map(({ id }, idx) => (
        <div
          key={id}
          id={keyboard[idx]}
          data-testid={keyboard[idx]}
          className={cn(
            styles.keyboardButton,
            { [styles.backspace]: keyboard[idx] === "backspace" },
            { [styles.enter]: keyboard[idx] === "enter" },
            {
              [styles["keyboardButton__green"]]:
                globalWordValidation[keyboard[idx]] === 1,
            },
            {
              [styles["keyboardButton__yellow"]]:
                globalWordValidation[keyboard[idx]] === 0,
            },
            {
              [styles["keyboardButton__black"]]:
                globalWordValidation[keyboard[idx]] === -1,
            }
          )}
          onClick={() => handleClick(keyboard[idx])}
        >
          {keyboard[idx] === "backspace" ? (
            <FontAwesomeIcon icon={faDeleteLeft} />
          ) : keyboard[idx] === "enter" ? (
            <FontAwesomeIcon icon={faArrowTurnDown} />
          ) : (
            keyboard[idx]
          )}
        </div>
      ))}
    </div>
  );
};
