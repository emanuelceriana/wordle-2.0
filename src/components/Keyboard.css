.keyboard {
  display: flex;
  flex-direction: row;
  justify-content: center;
  max-width: 38vw;
  flex-wrap: wrap;
}

.keyboard .keyboardButton {
  height: 3em;
  width: 3em;
  font-size: 1em;
  text-transform: capitalize;
  margin: 0.1em;
  border: 2px solid white;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #464347;
  cursor: pointer;
  position: relative;
  border-radius: 10px;
}

.keyboard .keyboardButton.green {
  background-color: rgba(43, 248, 43, 0.459) !important;
}

.keyboard .keyboardButton.yellow {
  background-color: rgba(255, 217, 0, 0.829) !important;
}

.keyboard .keyboardButton.black {
  background-color: rgba(0, 0, 0, 0.514) !important;
}

.keyboard .keyboardButton:hover {
  background-color: rgba(128, 128, 128, 0.74);
}

.keyboard .keyboardButton:nth-child(21) {
  flex-basis: 7.8em;
}

.keyboard .keyboardButton:nth-child(21) svg {
  rotate: 90deg;
}

.keyboard .keyboardButton:nth-child(28) {
  flex-basis: 5em;
}
