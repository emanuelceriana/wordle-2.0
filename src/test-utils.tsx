import React, { ReactNode } from "react";
import { render } from "@testing-library/react";
import { AppContext } from "./context/AppContext";

const MockProvider: React.FC<{
  children: ReactNode | ReactNode[];
  value: object;
}> = ({ children, value }) => (
  <AppContext.Provider value={value}>{children}</AppContext.Provider>
);

export function renderWithContext(element: React.ReactElement, value: object) {
  return render(<MockProvider value={value}>{element}</MockProvider>);
}

export function withContext(element: React.ReactElement, value: object) {
  return <MockProvider value={value}>{element}</MockProvider>;
}
