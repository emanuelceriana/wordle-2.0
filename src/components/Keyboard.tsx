import {
  faArrowTurnDown,
  faDeleteLeft,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import cn from "classnames";
import { GlobalWordArray } from "./Grid";
import "./Keyboard.css";
import { useCreateObjectList } from "./hooks/useCreateObjectList";
import { keyboardButtons } from "./constants";

interface KeyboardProps {
  globalWordArrayValidation: GlobalWordArray;
  activeRow: HTMLDivElement | null;
}

export const Keyboard = ({
  globalWordArrayValidation,
  activeRow,
}: KeyboardProps) => {
  const { objectList: keyboardButtonsIds } = useCreateObjectList({
    length: keyboardButtons.length,
  });

  const handleClick = (letter: string) => {
    const event = new KeyboardEvent("keydown", {
      key: letter,
      bubbles: true,
    });

    activeRow?.dispatchEvent(event);
  };

  return (
    <div className="keyboard">
      {keyboardButtonsIds.map(({ id }, idx) => (
        <div
          key={id}
          className={cn(
            "keyboardButton",
            { green: globalWordArrayValidation[keyboardButtons[idx]] === 1 },
            { yellow: globalWordArrayValidation[keyboardButtons[idx]] === 0 },
            { black: globalWordArrayValidation[keyboardButtons[idx]] === -1 }
          )}
          onClick={() => handleClick(keyboardButtons[idx])}
        >
          {keyboardButtons[idx] === "backspace" ? (
            <FontAwesomeIcon icon={faDeleteLeft} />
          ) : keyboardButtons[idx] === "enter" ? (
            <FontAwesomeIcon icon={faArrowTurnDown} />
          ) : (
            keyboardButtons[idx]
          )}
        </div>
      ))}
    </div>
  );
};
