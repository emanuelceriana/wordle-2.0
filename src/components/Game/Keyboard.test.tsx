import React from "react";
import { fireEvent, act } from "@testing-library/react";
import { Keyboard } from "./Keyboard";
import { renderWithContext } from "../../test-utils";

interface KeyboardProps {
  rowRefs: React.MutableRefObject<HTMLDivElement[]>;
  keyboardPreview: string[];
}

interface MockValueContext {
  globalWordValidation: {};
  activeRowIdx: number;
}

describe("<Keyboard />", () => {
  let keyboardProps: KeyboardProps = {
    rowRefs: { current: [document.createElement("div")] },
    keyboardPreview: [],
  };

  const mockValueContext: MockValueContext = {
    globalWordValidation: {},
    activeRowIdx: 0,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(
      <Keyboard {...keyboardProps} />,
      mockValueContext
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should dispatch keyup event when click keyboard letter", () => {
    const activeRowIdx = mockValueContext.activeRowIdx;
    const rowRefs = { current: [document.createElement("div")] };
    keyboardProps.rowRefs = rowRefs;

    rowRefs.current[activeRowIdx].onkeyup = jest.fn();

    const { getByTestId } = renderWithContext(
      <Keyboard {...keyboardProps} />,
      mockValueContext
    );

    const keyboardButtonA = getByTestId("a");

    act(() => {
      fireEvent.click(keyboardButtonA);
    });

    expect(rowRefs.current[activeRowIdx].onkeyup).toHaveBeenCalledTimes(1);
  });
});
