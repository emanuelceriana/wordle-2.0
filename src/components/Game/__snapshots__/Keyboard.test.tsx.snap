// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Keyboard /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="keyboard"
  >
    <div
      class="keyboardButton"
      data-testid="q"
      id="q"
    >
      q
    </div>
    <div
      class="keyboardButton"
      data-testid="w"
      id="w"
    >
      w
    </div>
    <div
      class="keyboardButton"
      data-testid="e"
      id="e"
    >
      e
    </div>
    <div
      class="keyboardButton"
      data-testid="r"
      id="r"
    >
      r
    </div>
    <div
      class="keyboardButton"
      data-testid="t"
      id="t"
    >
      t
    </div>
    <div
      class="keyboardButton"
      data-testid="y"
      id="y"
    >
      y
    </div>
    <div
      class="keyboardButton"
      data-testid="u"
      id="u"
    >
      u
    </div>
    <div
      class="keyboardButton"
      data-testid="i"
      id="i"
    >
      i
    </div>
    <div
      class="keyboardButton"
      data-testid="o"
      id="o"
    >
      o
    </div>
    <div
      class="keyboardButton"
      data-testid="p"
      id="p"
    >
      p
    </div>
    <div
      class="keyboardButton"
      data-testid="a"
      id="a"
    >
      a
    </div>
    <div
      class="keyboardButton"
      data-testid="s"
      id="s"
    >
      s
    </div>
    <div
      class="keyboardButton"
      data-testid="d"
      id="d"
    >
      d
    </div>
    <div
      class="keyboardButton"
      data-testid="f"
      id="f"
    >
      f
    </div>
    <div
      class="keyboardButton"
      data-testid="g"
      id="g"
    >
      g
    </div>
    <div
      class="keyboardButton"
      data-testid="h"
      id="h"
    >
      h
    </div>
    <div
      class="keyboardButton"
      data-testid="j"
      id="j"
    >
      j
    </div>
    <div
      class="keyboardButton"
      data-testid="k"
      id="k"
    >
      k
    </div>
    <div
      class="keyboardButton"
      data-testid="l"
      id="l"
    >
      l
    </div>
    <div
      class="keyboardButton"
      data-testid="z"
      id="z"
    >
      z
    </div>
    <div
      class="keyboardButton enter"
      data-testid="enter"
      id="enter"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrow-turn-down "
        data-icon="arrow-turn-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 64C14.3 64 0 49.7 0 32S14.3 0 32 0l96 0c53 0 96 43 96 96l0 306.7 73.4-73.4c12.5-12.5 32.8-12.5 45.3 0s12.5 32.8 0 45.3l-128 128c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 402.7 160 96c0-17.7-14.3-32-32-32L32 64z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="keyboardButton"
      data-testid="x"
      id="x"
    >
      x
    </div>
    <div
      class="keyboardButton"
      data-testid="c"
      id="c"
    >
      c
    </div>
    <div
      class="keyboardButton"
      data-testid="v"
      id="v"
    >
      v
    </div>
    <div
      class="keyboardButton"
      data-testid="b"
      id="b"
    >
      b
    </div>
    <div
      class="keyboardButton"
      data-testid="n"
      id="n"
    >
      n
    </div>
    <div
      class="keyboardButton"
      data-testid="m"
      id="m"
    >
      m
    </div>
    <div
      class="keyboardButton backspace"
      data-testid="backspace"
      id="backspace"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-delete-left "
        data-icon="delete-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M576 128c0-35.3-28.7-64-64-64H205.3c-17 0-33.3 6.7-45.3 18.7L9.4 233.4c-6 6-9.4 14.1-9.4 22.6s3.4 16.6 9.4 22.6L160 429.3c12 12 28.3 18.7 45.3 18.7H512c35.3 0 64-28.7 64-64V128zM271 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
